### PR TITLE
Use uconstr instead of the ad-hoc casted_constr in the exact tactic.

### PR DESCRIPTION
**Pull request by ppedrot**
This was the last `legacy` genarg, and it has been removed in this PR.

Depends on:
- #15170
- #15178

Overlays:
- https://github.com/HoTT/HoTT/pull/1596 (backwards compatible)

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1774345084

File: test-suite/output/ErrorLocation_12774_1.out [hunk 1]
```diff
@@ -1,3 +1,3 @@
-File "./output/ErrorLocation_12774_1.v", line 2, characters 13-14:
+File "./output/ErrorLocation_12774_1.v", line 2, characters 7-14:
```
Comment (SkySkimmer) [inline test-suite/output/ErrorLocation_12774_1.out:1]: Losing the location precision isn't great. I guess uconstr loses the ast locations? Could that be avoided?

Comment (ppedrot) [reply test-suite/output/ErrorLocation_12774_1.out:1]: No, actually uconstr has access to the locations. Rather the problem is that the heuristic used by Tacinterp to chose the "most precise location" (whatever that means) behaves badly on this example. I personally don't understand that part of the code, but the criterion should be refined to behave correctly on higher-order arguments of tactics, which essentially what `uconstr` is.

Comment (SkySkimmer) [reply test-suite/output/ErrorLocation_12774_1.out:1]: Where's that code?

Comment (ppedrot) [reply test-suite/output/ErrorLocation_12774_1.out:1]: Stuff around `Tacinterp.catch_error_with_trace_loc` should be a good lead.

**Comment by SkySkimmer** [comment]
Looks like it's not running TC correctly.

**Comment by ppedrot** [comment]
@coqbot ci minimize

File: test-suite/output/ErrorLocation_12774_1.out (hunk 1 above)
Comment (ppedrot) [reply test-suite/output/ErrorLocation_12774_1.out:1]: Note that there is already the same problem with `refine` in master, where the precise error location is never highlighted. This is fairly annoying when writing complex dependent programs that span several screens, but I've managed to survive it so far.

**Comment by SkySkimmer** [comment]
Is there a reason this is `empty` instead of `ist.lfun`? 
https://github.com/coq/coq/blob/1a57401e2249e7c824d04afa0b3a169ee72c45fd/plugins/ltac/tacinterp.ml#L1082

**Comment by ppedrot** [comment]
Good question, I don't remember. But I wouldn't be surprised I had to do this due to random Ltac behaviours around variable binding.

The very fact we're classifying the arguments of the closure without enough information is already very suspicious. But unfortunately to do this properly we would need to hook the Ltac runtime into the pretyper. This is now possible thanks to the use of open-recursion, but it has yet to be done.

**Comment by ppedrot** [comment]
Replying to myself because I should turn my tongue seven times before answering: this was obviously wrong. The closure is the one computed when the uconstr is constructed, as if it were a function closure. Using the one at the time of the evaluation totally defeats the concept of a closure. As evidenced by the current code, we only need it to carry metadata flags like `poly`.

**Comment by SkySkimmer** [comment]
Using `ist.lfun` is what interp_casted_constr was doing (through interp_gen)
and if we replace exact by refine (refine uses type_uconstr) in the minimized bedrock example we get the same nonsense "pf not bound" error.

**Comment by ppedrot** [comment]
OK, but the fix is not to pass the ist at the time of evaluation. It should be properly storing that data when building the closure.

**Comment by ppedrot** [comment]
Here is a smaller example for the HoTT failure:
```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-q" "-noinit" "-indices-matter"  "-top" "Z" "-native-compiler" "no") -*- *)
Declare ML Module "ltac_plugin".
Global Set Default Proof Mode "Classic".

Reserved Notation "x -> y" (at level 99, right associativity, y at level 200).
Reserved Notation "x = y" (at level 70, no associativity).

Declare Scope type_scope.

Global Open Scope type_scope.

Global Set Universe Polymorphism.

Notation "A -> B" := (forall (_ : A), B) : type_scope.

Create HintDb typeclass_instances discriminated.

Global Unset Strict Universe Declaration.

Notation Type0 := Set.

Cumulative Inductive paths {A : Type} (a : A) : A -> Type := idpath : paths a a.

Notation "x = y" := (@paths _ x y) : type_scope.

Declare Scope mc_scope.
Global Open Scope mc_scope.

Generalizable Variables A B f g x y.

Class SgOp A := sg_op: A -> A -> A.
Class Plus A := plus: A -> A -> A.
Class Mult A := mult: A -> A -> A.
Class Negate A := negate: A -> A.
Typeclasses Transparent SgOp Plus Mult Negate.

Global Instance plus_is_sg_op `{f : Plus A} : SgOp A := f.
Global Instance mult_is_sg_op `{f : Mult A} : SgOp A := f.

Section upper_classes.
  Context (A : Type@{i}).

  Class IsGroup {Aop : SgOp A} := {}.

  Context {Aplus : Plus A} {Anegate : Negate A}.

  Class IsRing := { ring_group :> @IsGroup plus_is_sg_op }.

End upper_classes.

Axiom preserves_negate : forall {A : Type} {Anegate : Negate A} 
  {B : Type} {Aop0 : SgOp B} {Anegate0 : Negate B}
  {_ : IsGroup B} {f : forall _ : A, B} (x : A),
  paths (f (negate x)) (negate (f x)).

Record Group := {
  group_type : Type;
  group_inverse : Negate group_type;
}.

Coercion group_type : Group >-> Sortclass.
Global Existing Instances group_inverse.

Axiom Int : Type0.
Axiom int_negation : Int -> Int.

Definition abgroup_Z : Group.
Proof.
 refine (Build_Group Int int_negation); repeat split.
Defined.

Record CRing := {
  cring_type : Type;
  cring_plus : Plus cring_type;
  cring_mult : Mult cring_type;
  cring_negate : Negate cring_type;
  cring_isring : IsRing cring_type;
}.

Coercion cring_type : CRing >-> Sortclass.

Global Existing Instances cring_plus cring_mult cring_negate cring_isring.

Axiom cring_catamorphism_fun : forall (R : CRing) (z : abgroup_Z), R.

Lemma cring_catamorphism_fun_negate {R} x
  : cring_catamorphism_fun R (negate x) = negate (cring_catamorphism_fun R x).
Proof.
simple notypeclasses refine (preserves_negate _).
1-2: exact _.
Qed.
```

**Comment by ppedrot** [comment]
This looks like an issue with the way typeclasses are resolved. I am not sure that the behaviour in master is actually correct, but here is what happens:
- in master, we try to find an instance of `SgOp`, which is solved by `cring_mult`. Morally, this is wrong because this will make the second resolution fail, since we this would have required instead `cring_plus`. Then we look for `@IsGroup R ?Aop0`, and that's the weird thing. There is an evar there, but it should have been resolved by the first typeclass run to `cring_mult`. The evar makes the resolution succeed.
- with this patch we "correctly" get in the second run the wrong instantiated goal `@IsGroup R cring_mult`, which fails.

**Comment by SkySkimmer** [comment]
In master there is a weird interleaving (I debugged `simple notypeclasses refine (preserves_negate _); exact _` but `1,2: exact _` is probably the same)
- interp_casted_constr on `_` with goal `?X7 : SgOp R`, TC infers cring_mult, no sideeffects on pre-existing evars
- interp_casted_constr on `_` with goal `?X8 : IsGroup R ?X7`, TC infers `ring_group R cring_plus cring_isring` with side effect `?X7 := cring_plus`
- run Tactics.exact_no_check focused on ?X8 with the above term

So ?X7 is filled by unification during TC for the second `_` inference, and refine is never called on it.

In other words we `interp` arguments on both goals before running the actual tactics, and side effects of the 2nd interp close the 1st goal before we refine can run on it.

**Comment by ppedrot** [comment]
This looks wrong, the first run is essentially lost?

**Comment by SkySkimmer** [comment]
Yeah, discarded by https://github.com/coq/coq/blob/1a57401e2249e7c824d04afa0b3a169ee72c45fd/engine/ftactic.ml#L41-L46 I think

**Comment by SkySkimmer** [comment]
Since there is a user visible change there should probably be a changelog.

**Review by SkySkimmer — APPROVED** [review]

**Comment by ppedrot** [comment]
And now itauto has decided to be broken again? What's with the CI curse these days?

**Comment by SkySkimmer** [comment]
@coqbot merge now
